### PR TITLE
Enable passes to lower tm_tensor ops to Linalg ops

### PR DESCRIPTION
**Pull request by Hanumanth04**
Change to enable required passes to lower the `tm_tensor` dialect to the `Linalg` dialect as part of the `torch-to-tosaLinalg` pass. We have a `tm-tensor-to-loops` pass that is expected to lower the `tm_tensor` dialect to the `Linalg` dialect. However, this pass works (lowers tm_tensor dialect to linalg dialect) only if the types of the operands to the `tm_tensor` ops are of type `memref,` so we have to run another (`tm-tensor-bufferize`) pass that converts `tensors` to `memref` types.

### CODE REVIEW

File: lib/Dialect/TorchConversion/Transforms/Passes.cpp [hunk 1]
```diff
@@ -10,6 +10,8 @@
 #include "torch-mlir/Dialect/TorchConversion/Transforms/Passes.h"
 #include "mlir/Conversion/Passes.h"
 #include "mlir/Dialect/MemRef/Transforms/Passes.h"
+#include "mlir/Dialect/Bufferization/Transforms/Passes.h"
```
Comment (Hanumanth04) [inline lib/Dialect/TorchConversion/Transforms/Passes.cpp]: Remove these

File: lib/Dialect/TorchConversion/Transforms/VerifyTosaLinalgBackendContract.cpp [hunk 2]
```diff
@@ -84,9 +86,10 @@ class VerifyTosaLinalgBackendContractPass
     target.addDynamicallyLegalDialect<affine::AffineDialect>(opHasLegalTypes);
     target.addDynamicallyLegalDialect<cf::ControlFlowDialect>(opHasLegalTypes);
     target.addDynamicallyLegalDialect<scf::SCFDialect>(opHasLegalTypes);
-    target.addDynamicallyLegalDialect<ml_program::MLProgramDialect>(
-        opHasLegalTypes);
-
+    target.addDynamicallyLegalDialect<ml_program::MLProgramDialect>(opHasLegalTypes);
+    target.addLegalDialect<memref::MemRefDialect>();
+    target.addLegalDialect<bufferization::BufferizationDialect>();    
+
```
Comment (sahas3) [inline lib/Dialect/TorchConversion/Transforms/VerifyTosaLinalgBackendContract.cpp]: Use `addDynamicallyLegalDialect` to be consistent with the other dialect checks.

File: test/Conversion/TorchToTosa/torch-backend-to-tosa-backend-pipeline-tm-scan.mlir [hunk 3]
```diff
@@ -0,0 +1,45 @@
+// RUN: torch-mlir-opt -pass-pipeline='builtin.module(torch-backend-to-tosa-linalg-backend-pipeline)' -split-input-file -verify-diagnostics %s | FileCheck %s
```
Comment (sahas3) [inline test/Conversion/TorchToTosa/torch-backend-to-tosa-backend-pipeline-tm-scan.mlir]: There's already a file for testing our pass. We can move this to that file instead of adding another one.

File: test/Conversion/TorchToTosa/torch-backend-to-tosa-backend-pipeline-tm-scan.mlir [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+// CHECK:          return %[[VAL_12]], %[[VAL_11]] : tensor<1x512xi64>, tensor<1xi64>
+// CHECK:         }
+
+func.func @tm_scan(%arg0: tensor<1x512xi32>) -> (tensor<1x512xi64>, tensor<1xi64>) {
+    %c0_i64 = arith.constant 0 : i64
+    %0 = tosa.cast %arg0 : (tensor<1x512xi32>) -> tensor<1x512xi64>
+    %1 = tensor.empty() : tensor<1x512xi64>
+    %2 = linalg.fill ins(%c0_i64 : i64) outs(%1 : tensor<1x512xi64>) -> tensor<1x512xi64>
+    %3 = tensor.empty() : tensor<1xi64>
+    %4 = linalg.fill ins(%c0_i64 : i64) outs(%3 : tensor<1xi64>) -> tensor<1xi64>
```
Comment (sahas3) [inline test/Conversion/TorchToTosa/torch-backend-to-tosa-backend-pipeline-tm-scan.mlir]: Do we need all these? Can't the inputs to `tm_tensor.scan` be passed as the function input directly?

**Review by sahas3 — CHANGES_REQUESTED** [review]

File: test/Conversion/TorchToTosa/torch-backend-to-tosa-backend-pipeline-tm-scan.mlir (hunk 4 above)
Comment (Hanumanth04) [reply test/Conversion/TorchToTosa/torch-backend-to-tosa-backend-pipeline-tm-scan.mlir]: No. IR need not be this complicated. I have simplified the IR.

File: test/Conversion/TorchToTosa/torch-backend-to-tosa-backend-pipeline-tm-scan.mlir (hunk 3 above)
Comment (Hanumanth04) [reply test/Conversion/TorchToTosa/torch-backend-to-tosa-backend-pipeline-tm-scan.mlir]: Fixed

File: lib/Dialect/TorchConversion/Transforms/VerifyTosaLinalgBackendContract.cpp (hunk 2 above)
Comment (Hanumanth04) [reply lib/Dialect/TorchConversion/Transforms/VerifyTosaLinalgBackendContract.cpp]: Fixed

File: lib/Dialect/TorchConversion/Transforms/VerifyTosaLinalgBackendContract.cpp [hunk 5]
```diff
@@ -39,12 +41,19 @@ class VerifyTosaLinalgBackendContractPass
...
     TypeConverter converter;
-    converter.addConversion([](RankedTensorType type) -> Type {
-      if (BaseMemRefType::isValidElementType(type.getElementType()))
+    converter.addConversion([](Type type) -> Type {
+      auto elemTy = type;
+      if (isa<TensorType>(type))
+        elemTy = cast<TensorType>(type).getElementType();
+      if (isa<MemRefType>(type))
+        elemTy = cast<MemRefType>(type).getElementType();      
+      if (BaseMemRefType::isValidElementType(elemTy))
```
Comment (sahas3) [inline lib/Dialect/TorchConversion/Transforms/VerifyTosaLinalgBackendContract.cpp:52]: Why was this change necessary?

**Review by sahas3 — APPROVED** [review]

File: lib/Dialect/TorchConversion/Transforms/VerifyTosaLinalgBackendContract.cpp (hunk 5 above)
Comment (Hanumanth04) [reply lib/Dialect/TorchConversion/Transforms/VerifyTosaLinalgBackendContract.cpp:52]: This is needed because during the verification of the legalization of `MemRefTypes,` we would fail while getting the element type without this change
